store/stat_store.go

```diff
@@ -0,0 +1,442 @@
+	if lastPutTime.Equal(time.Time{}) {
+		ss.resetLastPut(ts, tp.Value.(uint64))
+		return nil
+	}
+
+	if ts.Before(lastPutTime) {
+		// Ignore TimePoints with Timestamps in the past
+		return fmt.Errorf("the provided timepoint has a timestamp in the past")
+	}
+
+	if ts.Equal(lastPutTime) {
+		// update lastPut with the new TimePoint
+		newVal := tp.Value.(uint64)
+		if newVal > ss.lastPut.max {
+			ss.lastPut.max = newVal
+		}
+		oldAvg := ss.lastPut.average
+		n := float64(ss.lastPut.actualCount)
+		ss.lastPut.average = (float64(newVal) + (n * oldAvg)) / (n + 1)
+		ss.lastPut.actualCount++
+		return nil
+	}
+
+	// new point is in the future, lastPut needs to be flushed to the statStore.
+
+	// Determine how many resolutions in the future the new point is at.
+	// The statStore always represents values up until 1 resolution from lastPut.
+	numRes := uint8(0)
+	curr := ts
+	for curr.After(ss.lastPut.stamp) {
+		curr = curr.Add(-ss.resolution)
+		numRes++
+	}
+
+	// Create a new bucket if the buffer is empty
+	if ss.buffer.Front() == nil {
+		ss.newBucket(numRes)
+		ss.resetLastPut(ts, tp.Value.(uint64))
+		for ss.tpCount > ss.windowDuration {
+			ss.rewind()
+		}
+		return nil
+	}
+
+	lastEntry := ss.buffer.Front().Value.(tpBucket)
+	lastAvg := ss.lastPut.average
+
+	// Place lastPut in the latest bucket if the difference from its average
+	// is less than epsilon
+	if uint64(math.Abs(float64(lastEntry.value)-lastAvg)) < ss.epsilon {
+		lastEntry.count += numRes
+		ss.tpCount += numRes
+		if ss.lastPut.max > lastEntry.max {
+			lastEntry.max = ss.lastPut.max
+			lastEntry.maxIdx = lastEntry.count - 1
+		}
+		// update in list
+		ss.buffer.Front().Value = lastEntry
+	} else {
+		// Create a new bucket
+		ss.newBucket(numRes)
+	}
+
+	// Delete the earliest represented TimePoints if the window is full
+	for ss.tpCount > ss.windowDuration {
+		ss.rewind()
+	}
+
+	ss.resetLastPut(ts, tp.Value.(uint64))
+	return nil
+}
+
+// resetLastPut initializes the lastPut field of the statStore from a time and a value.
+func (ss *statStore) resetLastPut(timestamp time.Time, value uint64) {
+	ss.lastPut.stamp = timestamp
+	ss.lastPut.actualCount = 1
+	ss.lastPut.average = float64(value)
+	ss.lastPut.max = value
+}
+
+// newBucket appends a new bucket to the statStore, using the values of lastPut.
+// newBucket should be called BEFORE resetting lastPut.
+// newBuckets are created by rounding up the lastPut average to the closest epsilon.
+// The numRes parameter is the number of resolutions that this bucket will hold.
+// numRes represents the number of resolutions from the newest TimePoint to the lastPut.
+func (ss *statStore) newBucket(numRes uint8) {
+	newEntry := tpBucket{
+		count:  numRes,
+		value:  ((uint64(ss.lastPut.average) / ss.epsilon) + 1) * ss.epsilon,
+		max:    ss.lastPut.max,
+		maxIdx: 0,
+	}
+	ss.buffer.PushFront(newEntry)
+	ss.tpCount += numRes
+
+	// If this was the first bucket, update ss.start
+	if ss.start.Equal(time.Time{}) {
```

Comment:
Actually, for strings you do `if s != ""`. Maybe `if ss.start != time.Time{}`? Although it's essentially the same.